### ISSUE TITLE
Solutions for "kryptonite streams"

### ISSUE THREAD

**Issue by ericxtang** [livepeer/go-livepeer#1725]
We have seen certain streams would cause a spike in transcoding sessions.  This, in combination of the way we limit resource overuse via session limits, significantly affects the platform's ability to scale under load. 

It's unclear why - so we need to investigate this issue.

My recommendations are:
1. Automatically record segments that cause transcoding failures or very long transcoding time.  Understand whether there is a bug in our transcoding software.
2. Improve our retry logic to protect the network from this type of problems.

**Comment by iameli** [comment]
> Automatically record segments that cause transcoding failures or very long transcoding time. Understand whether there is a bug in our transcoding software.

Specifically, I think Bs should automatically upload to something like a `-failObjectStore` in the event that many many orchestrators choke on a given stream.

**Comment by yondonfu** [comment]
[Kryptonite video sample](https://storage.googleapis.com/lp_testharness_assets/kryptonite.mp4).

Agreed that solutions for both 1 and 2 in the OP will be important for addressing this issue. 

For 2, we'll need to consider not only the retry behavior of B, but also the retry behavior of an HTTP push client that is using B. B already caps the # of retries (configurable, but defaults to 3). However, an HTTP push client can retry a segment with the same B or other Bs which would result in a kryptonite segment being continuously submitted to Os. This can result in the following scenario:

1. HTTP push client submits to B
2. B retries with 3 Os and returns an error to the push client
3. HTTP push client retries the same segment with the same B or another B it has access to
4. B retries with 3 Os and returns an error to the push client
5. Repeat

Example of what the logs look like:

<img width="1628" alt="Screen Shot 2021-01-15 at 9 06 17 AM" src="https://user-images.githubusercontent.com/5933273/104736534-1e56d600-5711-11eb-83c8-9967979a1a41.png">

One way to avoid the above scenario is to cap the # of retries on the push client for a segment. But, this assumes that you can control the retry behavior of the push client which might not always be possible. So, it might be worthwhile to also consider whether there is a way to somehow throttle a push client for cases where the retry logic on the push client cannot be updated.

The ability of kryptonite segments to affect *all* Os that a B is aware of may also signal the need for improved workload isolation ([shuffle sharding](https://aws.amazon.com/builders-library/workload-isolation-using-shuffle-sharding/) as one example) which would sacrifice some redundancy in favor of better resiliency by limiting the # of Os that a kryptonite segment could affect.

**Comment by jailuthra** [comment]
I was able to reproduce a stuck transcode session by sending in this kryptonite segment `7564951696a2d255accf.tempfile` in [this tar](https://storage.googleapis.com/docs.livepeer.live/eli/segment-debugging/kryptonite.tar.gz) through push-api to a local B.
Shoutout to @iameli for extracting these out! :smile: 

Here's the [goroutine kill log](https://gist.github.com/jailuthra/badbe86b903aacef0e44af17f0517bad)

This particular segment smells fishy as it has video metadata (like resolution and start-time) but *no video frames*. Somewhere in our transcoder pipeline this case isn't handled properly.

**Possible solution**

This is similar to the previous no-video-frame issue https://github.com/livepeer/lpms/issues/203. Those segments didn't even have the resolution and start-time related metadata in the bitstream, and only affected us if it was the first segment of a transcoding session (so we handled it by probing before we started a hardware transcoding session). But for this one we will need to get into the weeds of where exactly the assumption of a frame exists and how we can gracefully error out from there.

**Open Question**

I am not sure if this is the only type of "kryptonite segment" that we're seeing lately or if there are other segments that look wildly different but still cause similar symptoms.
